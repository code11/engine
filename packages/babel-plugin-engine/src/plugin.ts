import * as Babel from "@babel/core";
import { VariableDeclaratorVisitor } from "./visitors";

export type PluginConfig = {
  viewLibrary: string;
};

type pluginType = (babel: typeof Babel, state: PluginConfig) => {};

export const plugin: pluginType = (babel, state) => {
  return {
    name: "babel-plugin-engine",
    visitor: {
      VariableDeclarator: VariableDeclaratorVisitor(babel, state),
    },
  };
};
