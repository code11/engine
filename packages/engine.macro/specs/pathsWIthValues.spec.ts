import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "babel-plugin-macros";

// const macroFile = "'./build/macro/index.macro'";
const macroFile = "'@c11/engine.macro'";
const config = require("./babelConfig.json");

pluginTester({
  plugin,
  babelOptions: {
    filename: __filename,
    ...config,
  },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should keep values in paths": {
      only: true,
      code: `
        import { producer } from ${macroFile}
        const val = "321"
        const obj = {
          foo: {
            bar: "234"
          }
        }
        const a: producer = ({
          a1 = get.foo[123],
          a2 = get.foo[val],
          a3 = get.foo["val"].foo,
          a4 = get.foo[obj.foo.bar].baz,
          a5 = get.foo[prop.foo][Request.foo][param.bar].baz,
          a6 = get.foo[\`bam\${val}baz\`].foo,
        }) => {  }
      `,
      snapshot: true,
    },
  },
});
