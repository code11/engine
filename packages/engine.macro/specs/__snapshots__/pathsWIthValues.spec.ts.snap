// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin should keep values in paths: should keep values in paths 1`] = `

import { producer } from '@c11/engine.macro'
const val = "321"
const obj = {
  foo: {
    bar: "234"
  }
}
const a: producer = ({
  a1 = get.foo[123],
  a2 = get.foo[val],
  a3 = get.foo["val"].foo,
  a4 = get.foo[obj.foo.bar].baz,
  a5 = get.foo[prop.foo][Request.foo][param.bar].baz,
  a6 = get.foo[\`bam\${val}baz\`].foo,
}) => {  }

      ↓ ↓ ↓ ↓ ↓ ↓

"use strict";

var val = "321";
var obj = {
  foo: {
    bar: "234",
  },
};
var a = {
  props: {
    type: "STRUCT",
    value: {
      a1: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: "123",
          },
        ],
      },
      a2: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: val,
          },
        ],
      },
      a3: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: "val",
          },
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
      a4: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: obj.foo.bar,
          },
          {
            type: "CONST",
            value: "baz",
          },
        ],
      },
      a5: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "EXTERNAL",
            path: ["foo"],
          },
          {
            type: "CONST",
            value: Request.foo,
          },
          {
            type: "INVOKE",
            path: ["bar"],
          },
          {
            type: "CONST",
            value: "baz",
          },
        ],
      },
      a6: {
        type: "GET",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "CONST",
            value: "bam".concat(val, "baz"),
          },
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
    },
    meta: {},
  },
  fn: function fn(_ref) {
    var a1 = _ref.a1,
      a2 = _ref.a2,
      a3 = _ref.a3,
      a4 = _ref.a4,
      a5 = _ref.a5,
      a6 = _ref.a6;
  },
};


`;
